### DISCUSSION TITLE
Ensure cron job doesn't run at same time on two nodes

### DISCUSSION THREAD

**Discussion by MrPerun** [quartznet/quartznet#2301] [Q&A]
I have quartz with a cron job, triggering each minute, in microservice application that allows scaling. So multiple instances of the same microservice with same quartz configuration can run at same time.

I have following configuration:
```
 quartz.Properties["quartz.scheduler.instanceId"] = "AUTO";
 quartz.UsePersistentStore(s =>
                {
                    s.UseSqlServer(sqlServer =>
                    {
                        sqlServer.ConnectionString = "..."
                    });

                    s.UseNewtonsoftJsonSerializer();
                    s.UseClustering();
                });
 quartz.AddJob<QuartzScheduledJobInvoker<TJob>>(opts =>
        {
            var jobKey = new JobKey("job1", "service1");
            opts.WithIdentity(jobKey);
            opts.DisallowConcurrentExecution();
        });
```
The job runs for exactly for 1min30s.
But when I run 2 instances, each minute one new instance is started. So there is overlap - in instance2 the job is started, while the previous one is still running on instance1.

Does Quartz have support to prevent this case or should I implement some kind of distributed lock myslef?

Thanks for hints.

**Comment by lahma** [comment]
You need to have database persistence (clustered setup) and proper configuration https://www.quartz-scheduler.net/documentation/faq.html#how-do-i-keep-a-job-from-firing-concurrently .

**Reply by MrPerun** [reply]
I have the persistence, hopefully correctly configured?

```
                quartz.UsePersistentStore(s =>
                {
                    s.UseSqlServer(sqlServer =>
                    {
                        sqlServer.ConnectionString = "..."
                    });

                    s.UseNewtonsoftJsonSerializer();
                    s.UseClustering();
                });
```

**Reply by lahma** [reply]
So you were just missing the `DisallowConcurrentExecution` mentioned in the FAQ?

**Reply by MrPerun** [reply]
Well, placing the `[DisallowConcurrentExecution]` attribute to job class helps.

I expected that configuring by fluent does the same, but it apparently doesn't:
```
quartz.AddJob<...>(opts => { opts.DisallowConcurrentExecution(); });
```

Could it be a bug?

The fluent would be better for my purpose, static annotation by attribute is quite limiting to me.

**Reply by lahma** [reply]
So writing the attribute in code is more limiting than writing the configuration in code?

**Reply by MrPerun** [reply]
Yes, in fact, I'm building small abstraction layer around quartz. The question is if it is good design , but doing so, it is limiting.

I have one quartz job class - in principle - not complete code:
```
class QuartzScheduledJobInvoker<TJob> : IJob
    where TJob : IMyInternalJob
{
    private readonly TJob job;

    public QuartzScheduledJobInvoker(TJob job)
    {
        this.job = job;
    }

    public async Task Execute(IJobExecutionContext context)
    {
            await job.RunMyLogic(cancellation);
    }
}
```

With this design, I must place the [DisallowConcurrentExecution] on the QuartzScheduledJobInvoker, which applies the behavior to all jobs in my app. And it is what I don't need/want.
